lib/pbench/server/api/__init__.py

```diff
@@ -57,7 +57,8 @@ def register_endpoints(api, app, config):
 
     api.add_resource(
         DatasetsContents,
-        f"{base_uri}/datasets/contents/<string:dataset>",
+        f"{base_uri}/datasets/contents/<string:dataset>/",
```

Comment:
I retract this... maybe... for now. This works great on my laptop venv with direct pytest, but doesn't seem to work under tox or in Jenkins. Something about the routing is still messed up and I haven't figured it out yet.

Reply:
Does you laptop have a newer/different version of Flask?

Reply:
Not flask version, or Python version. With a relatively low bandwidth during class, I didn't try to isolate every module difference. Instead I did some research and experimentation, and worked out a different approach. For one thing, my filtering of the URI pattern to generate the API key name (e.g., `datasets_contents`) falls prey to this quirk, and I've changed the `add_resource` calls to specify the desired tag name using the `endpoint` parameter. Then I made the endpoints generator smarter so it handles these alternatives... although I'm not confident of the *order* I'll see the `Rule` objects so I tried to make it resilient. Siddardh's case is "fancier" than mine, so I realize that I need another tweak, which I should probably do as a separate PR anyway...

Reply:
@dbutenhof, I saw that your tests are now passing!  🎉 

> I've changed the `add_resource` calls

Yeah, I saw; I posted questions about that in your PR.

> Siddardh's case is "fancier" than mine, so I realize that I need another tweak, which I should probably do as a separate PR anyway...

Maybe do it as a PR to this one (assuming that your PR merges before this one), and Siddardh can merge it in here and it can get into `main` that way?

Reply:
That's why I separated out #2956 ... ideally that'll go through quickly, and we'll both merge against it for the win.

Reply:
Yes, that was a good move.  (It would have been better if you'd done it even sooner.... 😉)

Reply:
"Sooner" is a relative term... I worked out the solution in #2956 yesterday during class, so I couldn't possibly have separated it out from #2949 much sooner than yesterday afternoon! 😆

Reply:
In this case, "sooner" would have been "before you added it to #2949".  🙂

Reply:
I glanced through this again and, just to be thorough, after #2949, the pattern here should resemble what I have for `ServerConfiguration`:

```
    api.add_resource(
        DatasetsContents,
        f"{base_uri}/datasets/contents/<string:dataset>",
        f"{base_uri}/datasets/contents/<string:dataset>/<path:path>",
        endpoint="datasets_contents",
        resource_class_args=(config, logger),
    )
```

You should be able to add a third pattern for `f"{base_uri}/datasets/contents/<string:dataset>/"` and have them all merged by the endpoint name.

Reply:
@dbutenhof, this conversation is currently marked "resolved"...if you're making a specific request, should you "unresolve" it?  (Will @siddardh-ra see and note your comment if the conversation is resolved?)

Also, you said, "after #2949" -- did you mean #2956?  (This PR doesn't need to wait until #2949 has been merged, does it?)

Reply:
Yeah @dbutenhof, noted. I've already updated the code once https://github.com/distributed-system-analysis/pbench/pull/2956 is merged. I will push my changes shortly with those updated changes :+1:

Reply:
Perhaps that was a bit ambiguous. #2956 changes the `Rule` parsing to enable both #2954 and #2949 to express these multiple endpoints without causing confusion. The intent of the note here was that #2954 should follow the example of (not yet merged) #2949 to leverage the change in #2956 rather than the earlier advice I'd given here. 😁